File: copilot-demo-labs/use-cases/09-technical-debt-refactoring/src/billing-engine.py

from datetime import datetime


TAX_RATES = {
    "IT": 0.22,
    "DE": 0.19,
    "FR": 0.20,
    "US": 0.0,
}

VIP_DISCOUNT = 0.05


def get_tax_rate(country: str) -> float:
    return TAX_RATES.get(country, 0.10)


def compute_line_total(line: dict) -> float:
    gross = line["price"] * line["qty"]
    discount = gross * (line.get("discountPercent", 0) / 100) if line.get("discountPercent") else 0
    return round(gross - discount, 2)


def calculate_invoice(customer: dict, lines: list[dict], apply_credit: bool = True) -> dict:
    subtotal = 0.0
    for line in lines:
        subtotal += compute_line_total(line)

    discount = subtotal * VIP_DISCOUNT if customer["isVip"] else 0
    taxable_amount = subtotal - discount
    tax_rate = get_tax_rate(customer["country"])
    tax = round(taxable_amount * tax_rate, 2)

    total = taxable_amount + tax

    credit_applied = 0.0
    if apply_credit and customer["creditBalance"] > 0:
        credit_applied = min(customer["creditBalance"], total)
        total = round(total - credit_applied, 2)

    return {
        "customerId": customer["id"],
        "lines": lines,
        "subtotal": subtotal,
        "discount": discount,
        "taxRate": tax_rate,
        "tax": tax,
        "creditApplied": credit_applied,
        "total": total,
        "currency": "EUR",
        "issuedAt": datetime.now(),
    }


def format_invoice_summary(invoice: dict) -> str:
    header = f"Invoice for {invoice['customerId']} - {invoice['issuedAt'].isoformat()[:10]}"
    body = "\n".join(
        f"  {line['description']}: {line['qty']} x EUR{line['price']}"
        for line in invoice["lines"]
    )

    totals = [
        f"  Subtotal: EUR{invoice['subtotal']:.2f}",
        f"  VIP discount: -EUR{invoice['discount']:.2f}" if invoice["discount"] > 0 else None,
        f"  Tax ({invoice['taxRate'] * 100:.0f}%): EUR{invoice['tax']:.2f}",
        f"  Credit applied: -EUR{invoice['creditApplied']:.2f}" if invoice["creditApplied"] > 0 else None,
        f"  Total: EUR{invoice['total']:.2f}",
    ]

    return f"{header}\n{body}\n{"\n".join(value for value in totals if value)}"

File: copilot-demo-labs/use-cases/09-technical-debt-refactoring/src/billing-engine.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Demo;

public sealed class InvoiceLine
{
    public required string Description { get; init; }
    public decimal Price { get; init; }
    public decimal Qty { get; init; }
    public decimal? DiscountPercent { get; init; }
}

public sealed class Customer
{
    public required string Id { get; init; }
    public required string Name { get; init; }
    public required string Country { get; init; }
    public bool IsVip { get; init; }
    public decimal CreditBalance { get; init; }
}

public sealed class Invoice
{
    public required string CustomerId { get; init; }
    public required List<InvoiceLine> Lines { get; init; }
    public decimal Subtotal { get; init; }
    public decimal Discount { get; init; }
    public decimal TaxRate { get; init; }
    public decimal Tax { get; init; }
    public decimal CreditApplied { get; init; }
    public decimal Total { get; init; }
    public required string Currency { get; init; }
    public DateTime IssuedAt { get; init; }
}

public static class BillingEngine
{
    private static readonly Dictionary<string, decimal> TaxRates = new()
    {
        ["IT"] = 0.22m,
        ["DE"] = 0.19m,
        ["FR"] = 0.20m,
        ["US"] = 0.0m,
    };

    private const decimal VipDiscount = 0.05m;

    public static decimal GetTaxRate(string country)
    {
        return TaxRates.TryGetValue(country, out var value) ? value : 0.10m;
    }

    public static decimal ComputeLineTotal(InvoiceLine line)
    {
        var gross = line.Price * line.Qty;
        var discount = line.DiscountPercent.HasValue
            ? gross * (line.DiscountPercent.Value / 100)
            : 0;
        return Math.Round(gross - discount, 2);
    }

    public static Invoice CalculateInvoice(Customer customer, List<InvoiceLine> lines, bool applyCredit = true)
    {
        var subtotal = 0m;
        foreach (var line in lines)
        {
            subtotal += ComputeLineTotal(line);
        }

        var discount = customer.IsVip ? subtotal * VipDiscount : 0;
        var taxableAmount = subtotal - discount;
        var taxRate = GetTaxRate(customer.Country);
        var tax = Math.Round(taxableAmount * taxRate, 2);

        var total = taxableAmount + tax;

        var creditApplied = 0m;
        if (applyCredit && customer.CreditBalance > 0)
        {
            creditApplied = Math.Min(customer.CreditBalance, total);
            total = Math.Round(total - creditApplied, 2);
        }

        return new Invoice
        {
            CustomerId = customer.Id,
            Lines = lines,
            Subtotal = subtotal,
            Discount = discount,
            TaxRate = taxRate,
            Tax = tax,
            CreditApplied = creditApplied,
            Total = total,
            Currency = "EUR",
            IssuedAt = DateTime.Now,
        };
    }

    public static string FormatInvoiceSummary(Invoice invoice)
    {
        var header = $"Invoice for {invoice.CustomerId} - {invoice.IssuedAt:yyyy-MM-dd}";
        var body = string.Join("\n", invoice.Lines.Select(line => $"  {line.Description}: {line.Qty} x EUR{line.Price}"));
        var totals = new List<string>
        {
            $"  Subtotal: EUR{invoice.Subtotal:F2}",
            $"  Tax ({invoice.TaxRate * 100:F0}%): EUR{invoice.Tax:F2}",
            $"  Total: EUR{invoice.Total:F2}",
        };

        if (invoice.Discount > 0)
        {
            totals.Insert(1, $"  VIP discount: -EUR{invoice.Discount:F2}");
        }

        if (invoice.CreditApplied > 0)
        {
            totals.Insert(totals.Count - 1, $"  Credit applied: -EUR{invoice.CreditApplied:F2}");
        }

        return $"{header}\n{body}\n{string.Join("\n", totals)}";
    }
}
